core/domain/learner_progress_services.py

```diff
@@ -447,41 +544,49 @@ def get_all_completed_collection_ids(user_id):
         return []
 
 
-def get_completed_collection_summaries(user_id):
+def get_filtered_completed_collection_summaries(user_id, collection_summaries,
+                                                collection_ids):
     """Returns a list of summaries of the completed collection ids, the
     number of collections deleted from the list as they are no longer present
     and the number of collections being shifted to the incomplete section on
     account of new addition of explorations.
 
     Args:
         user_id: str. The id of the learner.
+        collection_summaries: list(CollectionSummary). The list of collection
+            summary domain objects to be filtered.
+        collection_ids: list(str). The ids of the collection corresponding to
+            the collection summary domain objects.
 
     Returns:
-        list(CollectionSummary). A list with the summary domain objects of the
-            completed collections.
-        int. The number of explorations deleted from the list as they are no
-            longer present.
-        int. The number of collections moved to the incomplete section on
+        tuple. A 3-tuple whose elements are as follows:
+        - list(CollectionSummary). A filtered list with the summary domain
+            objects of the completed collections.
+        - list(str). The ids of the collections that are no longer present.
+        - list(CollectionSummary). The summaries corresponding to those
+            collections which have been moved to the in progress section on
             account of new explorations being added to them.
     """
-    completed_collection_ids = get_all_completed_collection_ids(user_id)
-
-    number_deleted = 0
+    index = 0
+    nonexistent_completed_collection_ids = []
     completed_to_incomplete_collections = []
-    for collection_id in completed_collection_ids:
-        if not collection_services.does_collection_exists(collection_id):
-            number_deleted = number_deleted + 1
-            remove_collection_from_completed_list(user_id, collection_id)
+    filtered_completed_collection_summaries = []
+    for collection_summary in collection_summaries:
+        if collection_summary is None:
+            nonexistent_completed_collection_ids.append(collection_ids[index])
         elif collection_services.get_next_exploration_ids_to_complete_by_user(
```

Comment:
Oh dear. There are multiple datastore get() call hiding here (take a look at the function in collection_services), and that means you'll be iterating over all a user's collections and fetching each individually, every single time the learner dashboard is loaded.

I suggest that you fetch all the collections using a get_multi(), pass completed_exploration_ids into this function, and use something like collection.get_next_exploration_ids(completed_exploration_ids) instead.